styx-service-common/src/test/java/com/spotify/styx/storage/DatastoreStorageTest.java

```diff
@@ -768,159 +796,140 @@ private Workflow workflow(WorkflowId workflowId) {
-    final DatastoreStorage storage = new DatastoreStorage(datastore, Duration.ZERO, storageTransactionFactory, executor);
+  public void runInTransactionWithRetriesShouldCallFunction() throws Exception {
+    final DatastoreStorage storage = new DatastoreStorage(datastore, storageTransactionFactory,
+        executor, logger);
     final CheckedDatastoreTransaction transaction = datastore.newTransaction();
     final DatastoreStorageTransaction storageTransaction = spy(new DatastoreStorageTransaction(transaction));
     when(storageTransactionFactory.apply(any())).thenReturn(storageTransaction);
 
     when(transactionFunction.apply(any())).thenReturn("foo");
 
-    String result = storage.runInTransaction(transactionFunction);
+    String result = storage.runInTransactionWithRetries(transactionFunction);
 
     assertThat(result, is("foo"));
     verify(transactionFunction).apply(storageTransaction);
-    verify(storageTransaction).commit();
-    verify(storageTransaction, never()).rollback();
   }
 
   @Test
-  public void runInTransactionShouldCallFunctionAndRollbackOnFailure() throws Exception {
-    final DatastoreStorage storage = new DatastoreStorage(datastore, Duration.ZERO, storageTransactionFactory, executor);
-    final CheckedDatastoreTransaction transaction = datastore.newTransaction();
-    final DatastoreStorageTransaction storageTransaction = spy(new DatastoreStorageTransaction(transaction));
-    when(storageTransactionFactory.apply(any())).thenReturn(storageTransaction);
-
+  public void runInTransactionWithRetriesShouldPropagateUserException() throws Exception {
     final Exception expectedException = new FooException();
     when(transactionFunction.apply(any())).thenThrow(expectedException);
 
     try {
-      storage.runInTransaction(transactionFunction);
+      storage.runInTransactionWithRetries(transactionFunction);
       fail("Expected exception!");
     } catch (FooException e) {
       // Verify that we can throw a user defined checked exception type inside the transaction
       // body and catch it
       assertThat(e, is(expectedException));
     }
 
-    verify(transactionFunction).apply(storageTransaction);
-    verify(storageTransaction, never()).commit();
-    verify(storageTransaction).rollback();
+    verify(transactionFunction).apply(any());
   }
 
   @Test
-  public void runInTransactionShouldCallFunctionAndRollbackOnPreCommitConflict() throws Exception {
-    final DatastoreStorage storage = new DatastoreStorage(datastore, Duration.ZERO, storageTransactionFactory, executor);
-    final CheckedDatastoreTransaction transaction = datastore.newTransaction();
-    final DatastoreStorageTransaction storageTransaction = spy(new DatastoreStorageTransaction(transaction));
-    when(storageTransactionFactory.apply(any())).thenReturn(storageTransaction);
-
-    final Exception expectedException = new DatastoreException(10, "", "");
+  public void runInTransactionWithRetriesShouldRetry() throws Exception {
```

Comment:
It doesn't look like we actually verify that a retry happened, or am I missing something?

Reply:
I would guess this test is causing my confusion. The name looks like it would use the retry settings, but the implementation doesn't.

Reply:
Good catch. Currently the retry tests were in `DatastoreStorageTransactionTest.java`. Moving  them to `DatastoreStorageTest.java`.